### ISSUE TITLE
Error building GDAL 3.0.0 with latest dependent libs

### ISSUE THREAD

**Issue by NikkittaP** [OSGeo/gdal#1672]
I think I found a bug.
Firstly I was thinking that this is problem with VS2015 x64, but now I get the same error on Debian-based linux.

I am building GDAL 3.0.0 and some dependent libs completely from source. Here is the list and versions of sources used:

- zlib 1.2.11
- libpng 1.6.37
- libjpeg-turbo 2.0.2
- libtiff 4.0.10
- libcurl 7.65.1
- proj.4 6.1.0
- libgeotiff 1.5.1

**Windows build**
Yesterday I had a clean installation of Windows 10 x64 and VS2015.3.
To build all the libs I use CMake, MSBuild and nmake.

On both Release and debug builds of GDAL I get:
`C:\OSG\test_osg\build\gdal>if exist *.ilk del *.ilk 
	link /nologo /dll /INCLUDE:OSRValidate  /INCLUDE:OGR_G_GetPointCount  /INCLUDE:OGRRegisterAll /INCLUDE:GDALSimpleImageWarp  /INCLUDE:GDALReprojectImage  /INCLUDE:GDALComputeMedianCutPCT  /INCLUDE:GDALDitherRGB2PCT  /INCLUDE:OCTNewCoordinateTransformation port\*.obj gcore\*.obj alg\*.obj frmts\o\*.obj ogr\ogrsf_frmts\ogrsf_frmts.lib ogr\ogr.lib gnm\*.obj gnm\gnm_frmts\o\*.obj third_party\o\*.obj apps\commonutils.obj apps\gdalinfo_lib.obj apps\gdal_translate_lib.obj apps\gdalwarp_lib.obj apps\ogr2ogr_lib.obj  apps\gdaldem_lib.obj apps\nearblack_lib.obj apps\gdal_grid_lib.obj apps\gdal_rasterize_lib.obj apps\gdalbuildvrt_lib.obj                C:\OSG\test_osg\install\libjpeg\Release\lib\jpeg.lib    C:\OSG\test_osg\install\libgeotiff\Release\lib\geotiff.lib C:\OSG\test_osg\install\libtiff\Release\lib\tiff.lib C:\OSG\test_osg\install\proj\Release\lib\proj_6_1.lib            C:\OSG\test_osg\install\libcurl\Release\lib\libcurl_imp.lib wsock32.lib wldap32.lib winmm.lib    legacy_stdio_definitions.lib odbc32.lib odbccp32.lib user32.lib  C:\OSG\test_osg\install\libpng\Release\lib\libpng16.lib C:\OSG\test_osg\install\zlib\Release\lib\zlib.lib                         ws2_32.lib  kernel32.lib psapi.lib gcore\Version.res  /out:gdal300.dll /implib:gdal_i.lib   
   Creating library gdal_i.lib and object gdal_i.exp
gdal_crs.obj : error LNK2019: unresolved external symbol GDALSerializeGCPListToXML referenced in function GDALSerializeGCPTransformer
gdal_crs.obj : error LNK2019: unresolved external symbol GDALDeserializeGCPListFromXML referenced in function GDALDeserializeGCPTransformer
ogrsf_frmts.lib(ogrregisterall.obj) : error LNK2019: unresolved external symbol RegisterOGRGFT referenced in function OGRRegisterAllInternal
gdal300.dll : fatal error LNK1120: 3 unresolved externals`

**Linux build**
To build all the libs I use CMake and make.
To build GDAL 3.0.0 and use all my prebuilt libs I use the following commands:

`export LD_LIBRARY_PATH=/home/nikita/OSG/test_osg/install/zlib/Release/lib:/home/nikita/OSG/test_osg/install/libpng/Release/lib:/home/nikita/OSG/test_osg/install/libjpeg/Release/lib:/home/nikita/OSG/test_osg/install/libtiff/Release/lib:/home/nikita/OSG/test_osg/install/libcurl/Release/lib:/home/nikita/OSG/test_osg/install/proj/Release/lib env`
`ldconfig`
`export LDFLAGS=-Wl,-rpath,/home/nikita/OSG/test_osg/install/zlib/Release/lib,-rpath,/home/nikita/OSG/test_osg/install/libpng/Release/lib,-rpath,/home/nikita/OSG/test_osg/install/libjpeg/Release/lib,-rpath,/home/nikita/OSG/test_osg/install/libtiff/Release/lib,-rpath,/home/nikita/OSG/test_osg/install/libcurl/Release/lib,-rpath,/home/nikita/OSG/test_osg/install/proj/Release/lib
sh ./configure --prefix=/home/nikita/OSG/test_osg/install/gdal/Release --with-libz=/home/nikita/OSG/test_osg/install/zlib/Release --with-png=/home/nikita/OSG/test_osg/install/libpng/Release --with-jpeg=/home/nikita/OSG/test_osg/install/libjpeg/Release --with-libtiff=/home/nikita/OSG/test_osg/install/libtiff/Release --with-geotiff=/home/nikita/OSG/test_osg/install/libgeotiff/Release --with-curl=/home/nikita/OSG/test_osg/install/libcurl/Release/bin/curl-config --with-proj=/home/nikita/OSG/test_osg/install/proj/Release`
`make -j6`

And I get the following error, smae as on Windows:
`/home/nikita/OSG/test_osg/build/gdal/.libs/libgdal.so: undefined reference to 'GDALSerializeGCPListToXML'
/home/nikita/OSG/test_osg/build/gdal/.libs/libgdal.so: undefined reference to 'GDALDeserializeGCPListFromXML'
collect2: error: ld returned 1 exit status
make[1]: *** [gdalinfo] Error 1
make: *** [apps-target] Error 2`

All libs are latest stable versions.
Both systems and builds are x64.

**Comment by rouault** [comment]
Hum, this is a weird one. I can't see how the external libs would influence GDALDeserializeGCPListFromXML() being compiled or not (it is defined in gcore/gdal_misc.cpp unconditionnaly of any external lib). Can you retry for example the Linux build without any external dependency (make sure you "make clean" between all build attempts), except for PROJ 6.1 which is the only one require ?  Hopefully that will succeed, and then you'll have to re-add the lib one by one to pinpoint the one that causes the issue.

**Comment by NikkittaP** [comment]
Oh, I'm truly sorry, it seems that there was some kind of problem with the files of GDAL sources in my git. I had some additional files in the GDAL sources, like there should be only FILE.cpp and I had also FILE.c. And that was the problem. I cleaned everything and now it builds on both systems.

Thank you very much for your suggestion - I used it to build GDAL over and over adding each library, and finally I realized that the problem was in my sources of GDAL.
